wcfsetup/install/files/lib/system/form/builder/field/UploadFormField.class.php

```diff
@@ -296,6 +304,41 @@ public function validate()
                 }
             }
         }
+
+        if ($this->allowedFileExtensions !== null) {
+            // File extensions that contain a dot must be checked otherwise,
+            // the function `UploadFile::getFilenameExtension()` only returns the last part of the file extension.
+            //
+            // For example, if the file is called `blub.tar.gz`, only `gz` is returned.
+            // In this case, we check whether the file ends with `.tar.gz`.
+            $specialFileExtensions = \array_filter($this->allowedFileExtensions, function ($extension) {
+                return \str_contains($extension, '.');
+            });
+            $allowedFileExtensions = \array_filter($this->allowedFileExtensions, function ($extension) {
+                return !\str_contains($extension, '.');
+            });
+
+            foreach ($this->getValue() as $file) {
+                if (!\in_array($file->getFilenameExtension(), $allowedFileExtensions)) {
```

Comment:
Please invert this condition to `continue` early. This will reduce the nesting level and avoid the use of `continue 2` below.